app/components/secure/AuthWalletKeys.tsx

```diff
@@ -27,50 +30,105 @@ export type AuthWalletKeysType = {
     authenticateWithPasscode: (style?: AuthStyle) => Promise<WalletKeys>,
 }
 
-function loadPasscodeState(address: Address, isTestnet: boolean) {
-    return storage.getString(`${address.toFriendly({ testOnly: isTestnet })}/${passcodeStateKey}`);
+export function getPasscodeState(address: Address, isTestnet: boolean) {
+    return (storage.getString(`${address.toFriendly({ testOnly: isTestnet })}/${passcodeStateKey}`) ?? null) as PasscodeState | null;
 }
 
 export const AuthWalletKeysContext = React.createContext<AuthWalletKeysType | null>(null);
 
 export const AuthWalletKeysContextProvider = React.memo((props: { children?: any }) => {
     const engine = useEngine();
+    const safeAreaInsets = useSafeAreaInsets();
     const { Theme, AppConfig } = useAppConfig();
     const [auth, setAuth] = useState<AuthProps | null>(null);
 
     const authenticate = useCallback(async (style?: AuthStyle) => {
+
+        // Reject previous auth promise
         if (auth) {
             auth.promise.reject();
         }
+
+        // Clear previous auth
         setAuth(null);
+
         const acc = getCurrentAddress();
-        // Try biometric auth then fallback to passcode if fails
+
+        // Check if migrated to new Passcode and Biometrics keys system
+        const migrated = getBiometricsMigrated(AppConfig.isTestnet);
+
         try {
```

Comment:
Refactor to remove nested try/catch statements

Reply:
done